File: solutions/cpp/Computing_x^n.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <iostream>
#include <cassert>
#include <limits>
#include <list>
#include <queue>
#include <random>
#include <stdexcept>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::invalid_argument;
using std::list;
using std::queue;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

// @include
list<int> GetMinimumExpression(int n) {
  if (n == 1) {
    return {1};
  }

  queue<list<int>> exp_lists;
  // Constructs the initial list with one node whose value is 1.
  exp_lists.emplace(1, 1);
  while (exp_lists.empty() == false) {
    list<int> exp = exp_lists.front();
    exp_lists.pop();
    // Tries all possible combinations in list exp.
    for (const int& a : exp) {
      int sum = a + exp.back();
      if (sum > n) {
        break;  // No possible solution.
      }

      list<int> new_exp(exp);
      new_exp.emplace_back(sum);
      if (sum == n) {
        return new_exp;
      }
      exp_lists.emplace(new_exp);
    }
  }
  // @exclude
  throw invalid_argument("unknown error");  // This line should never be called.
  // @include
}
// @exclude

void small_test() {
  auto res = GetMinimumExpression(88);
  list<int> golden_res = {1, 2, 3, 4, 7, 11, 22, 44, 88};
  assert(res.size() == golden_res.size());
  assert(equal(res.begin(), res.end(), golden_res.begin()));
  res = GetMinimumExpression(67);
  golden_res = {1, 2, 3, 4, 8, 16, 32, 35, 67};
  assert(res.size() == golden_res.size());
  assert(equal(res.begin(), res.end(), golden_res.begin()));
}

int main(int argc, char* argv[]) {
  small_test();
  default_random_engine gen((random_device())());
  int n;
  if (argc == 2) {
    n = atoi(argv[1]);
  } else {
    uniform_int_distribution<int> n_dis(1, 100);
    n = n_dis(gen);
  }
  cout << "n = " << n << endl;
  list<int> min_exp = GetMinimumExpression(n);
  for (const int& t : min_exp) {
    cout << t << ' ';
  }
  cout << endl;
  cout << "size = " << min_exp.size() << endl;
  return 0;
}

File: static/python/computing_x^n.py
# Computing_x^n.cpp b4b3a70d8ab942579f85b4416f980d05831af969
import sys
import random
import collections


# @include
def get_shortest_straight_line_program(n):
    if n == 1:
        return [1]

    SLPs = collections.deque()  # SLP is acronym for straight line program.
    # Constructs the initial SLP with one node whose value is 1.
    SLPs.append([1])
    while SLPs:
        candidate_SLP = SLPs.popleft()
        # Tries all possible combinations in candidate_SLP.
        for a in candidate_SLP:
            power = a + candidate_SLP[-1]
            if power > n:
                break  # No possible solution for candidate_SLP.
            new_SLP = candidate_SLP.copy()
            new_SLP.append(power)
            if power == n:
                return new_SLP
            SLPs.append(new_SLP)

# @exclude
    raise ValueError('unknown error')  # This line should never be called.


def small_test():
    res = get_shortest_straight_line_program(88)
    golden_res = [1, 2, 3, 4, 7, 11, 22, 44, 88]
    assert res == golden_res
    res = get_shortest_straight_line_program(67)
    golden_res = [1, 2, 3, 4, 8, 16, 32, 35, 67]
    assert res == golden_res


def main():
    small_test()
    if len(sys.argv) == 2:
        n = int(sys.argv[1])
    else:
        n = random.randint(1, 100)
    print('n =', n)
    min_exp = get_shortest_straight_line_program(n)
    print(*min_exp)
    print('size =', len(min_exp))


if __name__ == '__main__':
    main()
